pkg/joinserver/joinserver.go

```diff
@@ -322,19 +333,29 @@ func (js *JoinServer) HandleJoin(ctx context.Context, req *ttnpb.JoinRequest, au
 			if req.SelectedMacVersion.IncrementDevNonce() {
 				if (dn != 0 || dev.LastDevNonce != 0 || dev.LastJoinNonce != 0) && !dev.ResetsJoinNonces {
 					if dn <= dev.LastDevNonce {
+						registerDevNonceTooSmall(ctx, req)
 						return nil, nil, errDevNonceTooSmall.New()
 					}
 				}
 				dev.LastDevNonce = dn
 				paths = append(paths, "last_dev_nonce")
 			} else {
-				i := sort.Search(len(dev.UsedDevNonces), func(i int) bool { return dev.UsedDevNonces[i] >= dn })
-				if i >= len(dev.UsedDevNonces) || dev.UsedDevNonces[i] != dn {
-					dev.UsedDevNonces = append(dev.UsedDevNonces, 0)
-					copy(dev.UsedDevNonces[i+1:], dev.UsedDevNonces[i:])
-					dev.UsedDevNonces[i] = dn
+				isReuse := false
+				for _, v := range dev.UsedDevNonces {
+					if v == dn {
+						isReuse = true
+						break
+					}
+				}
```

Comment:
```suggestion
				for i := len(dev.UsedDevNonces) - 1; i >= 0; i-- {
					if dev.UsedDevNonces[i] == dn {
						isReuse = true
						break
					}
				}
```

Reply:
True, it's more likely for the reused DevNonce to be in the start of the queue. Updated it.